python/setup.py

```diff
@@ -133,8 +143,68 @@ def run(self):
                       'bundle the Arrow C++ headers')] +
                     _build_ext.user_options)
 
+    def get_arrow_build_options(self):
```

Comment:
I have to say I find this solution quite horrible. Enabling or disabling PyArrow components manually is a bit cumbersome, but not terribly so. If there's no better way to query Arrow CMake arguments than parsing some adhoc output format from a hidden CMake option, then I would rather not have this.

@kou Any thoughts?

Reply:
We can do this in `python/CMakeLists.txt`.
We can refer all `ARROW_*` variables after `find_package(Arrow)`.
We can use these values as the default value of `PYARROW_BUILD_*` (Note that the following example isn't completed. It breaks at least `PYARROW_CXXFLAGS`.):

```diff
diff --git a/python/CMakeLists.txt b/python/CMakeLists.txt
index 29f8d2da7..a002b4e4e 100644
--- a/python/CMakeLists.txt
+++ b/python/CMakeLists.txt
@@ -108,25 +108,6 @@ if(UNIX)
   endif()
 endif()
 
-# Top level cmake dir
-if("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}")
-  option(PYARROW_BUILD_ACERO "Build the PyArrow Acero integration" OFF)
-  option(PYARROW_BUILD_CUDA "Build the PyArrow CUDA support" OFF)
-  option(PYARROW_BUILD_DATASET "Build the PyArrow Dataset integration" OFF)
-  option(PYARROW_BUILD_FLIGHT "Build the PyArrow Flight integration" OFF)
-  option(PYARROW_BUILD_GANDIVA "Build the PyArrow Gandiva integration" OFF)
-  option(PYARROW_BUILD_ORC "Build the PyArrow ORC integration" OFF)
-  option(PYARROW_BUILD_PARQUET "Build the PyArrow Parquet integration" OFF)
-  option(PYARROW_BUILD_PARQUET_ENCRYPTION
-         "Build the PyArrow Parquet encryption integration" OFF)
-  option(PYARROW_BUNDLE_ARROW_CPP "Bundle the Arrow C++ libraries" OFF)
-  option(PYARROW_BUNDLE_CYTHON_CPP "Bundle the C++ files generated by Cython" OFF)
-  option(PYARROW_GENERATE_COVERAGE "Build with Cython code coverage enabled" OFF)
-  set(PYARROW_CXXFLAGS
-      ""
-      CACHE STRING "Compiler flags to append when compiling Arrow")
-endif()
-
 find_program(CCACHE_FOUND ccache)
 if(CCACHE_FOUND
    AND NOT CMAKE_C_COMPILER_LAUNCHER
@@ -267,6 +248,25 @@ include(GNUInstallDirs)
 
 find_package(Arrow REQUIRED)
 
+# Top level cmake dir
+if("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}")
+  option(PYARROW_BUILD_ACERO "Build the PyArrow Acero integration" ${ARROW_ACERO})
+  option(PYARROW_BUILD_CUDA "Build the PyArrow CUDA support" ${ARROW_CUDA})
+  option(PYARROW_BUILD_DATASET "Build the PyArrow Dataset integration" ${ARROW_DATASET})
+  option(PYARROW_BUILD_FLIGHT "Build the PyArrow Flight integration" ${ARROW_FLIGHT})
+  option(PYARROW_BUILD_GANDIVA "Build the PyArrow Gandiva integration" ${ARROW_GANDIVA})
+  option(PYARROW_BUILD_ORC "Build the PyArrow ORC integration" ${ARROW_ORC})
+  option(PYARROW_BUILD_PARQUET "Build the PyArrow Parquet integration" ${ARROW_PARQUET})
+  option(PYARROW_BUILD_PARQUET_ENCRYPTION
+         "Build the PyArrow Parquet encryption integration" ${PARQUET_REQUIRE_ENCRYPTION})
+  option(PYARROW_BUNDLE_ARROW_CPP "Bundle the Arrow C++ libraries" OFF)
+  option(PYARROW_BUNDLE_CYTHON_CPP "Bundle the C++ files generated by Cython" OFF)
+  option(PYARROW_GENERATE_COVERAGE "Build with Cython code coverage enabled" OFF)
+  set(PYARROW_CXXFLAGS
+      ""
+      CACHE STRING "Compiler flags to append when compiling Arrow")
+endif()
+
 set(PYARROW_CPP_ROOT_DIR pyarrow/src)
 set(PYARROW_CPP_SOURCE_DIR ${PYARROW_CPP_ROOT_DIR}/arrow/python)
 set(PYARROW_CPP_SRCS
```

Reply:
@pitrou finally getting round to messing with this now c++ is in - I discover that there is actually an official way to query arrow options as JSON. I'll look at that next.

The reason for needing this for pyodide is that pyodide builds everything in an isolated environment which loses the environment variables that are passed in  to set these things up. It's either that or make a file to load settings in, but even if one supported a settings file it would be good to default to building a pyarrow with the same options as the linked in libarrow.

@kou - I think it has to be done before the cmake build is run because the value of the options alters some things in setup.py

Reply:
It seems that you want to enable PyArrow components if corresponding Arrow C++ components are enabled.
If so, you don't need to specify `PYARROW_BUILD_*` CMake variables by `setup.py` or can ignore CMake variables specified by`setup.py` in `CMakeLists.txt`. You can use https://github.com/apache/arrow/pull/37822/files#r1344887227 for this approach.

Reply:
FWIW this has been merged in the meantime, so you should be able to simplify this again now